src/rmem.c

```diff
@@ -0,0 +1,470 @@
+        mempool.stack.mem = malloc(1 + mempool.stack.size*sizeof *mempool.stack.mem);
+        if( mempool.stack.mem==NULL ) {
+            mempool.stack.size = 0UL;
+            return mempool;
+        } else {
+            mempool.stack.base = mempool.stack.mem + mempool.stack.size;
+            return mempool;
+        }
+    }
+}
+
+struct MemPool MemPool_FromBuffer(const size_t size, void *buf)
+{
+    struct MemPool mempool = {0};
+    if( size==0UL || buf==NULL || size<=sizeof(struct MemNode) )
+        return mempool;
+    else {
+        mempool.stack.size = size;
+        mempool.stack.mem = buf;
+        mempool.stack.base = mempool.stack.mem + mempool.stack.size;
+        return mempool;
+    }
+}
+
+void MemPool_Destroy(struct MemPool *const mempool)
+{
+    if( mempool==NULL || mempool->stack.mem==NULL )
+        return;
+    else {
+        free(mempool->stack.mem);
+        *mempool = (struct MemPool){0};
+    }
+}
+
+void *MemPool_Alloc(struct MemPool *const mempool, const size_t size)
+{
+    if( mempool==NULL || size==0UL || size > mempool->stack.size )
+        return NULL;
+    else {
+        struct MemNode *new_mem = NULL;
+        const size_t ALLOC_SIZE = size + sizeof *new_mem;
+        if( mempool->freeList.head != NULL ) {
+            const size_t MEM_SPLIT_THRESHOLD = sizeof(intptr_t);
+            // if the freelist is valid, let's allocate FROM the freelist then!
+            for( struct MemNode **inode = &mempool->freeList.head; *inode != NULL; inode = &(*inode)->next ) {
+                if( (*inode)->size < ALLOC_SIZE )
+                    continue;
+                else if( (*inode)->size <= ALLOC_SIZE + MEM_SPLIT_THRESHOLD ) {
+                    // close in size - reduce fragmentation by not splitting.
+                    new_mem = *inode;
+                    _RemoveNode(inode);
+                    mempool->freeList.len--;
+                    new_mem->next = new_mem->prev = NULL;
+                    break;
+                } else {
+                    // split the memory chunk.
+                    new_mem = (struct MemNode *)( (uint8_t *)*inode + ((*inode)->size - ALLOC_SIZE) );
+                    (*inode)->size -= ALLOC_SIZE;
+                    new_mem->size = ALLOC_SIZE;
+                    new_mem->next = new_mem->prev = NULL;
+                    break;
+                }
+            }
+        }
+        
+        if( new_mem==NULL ) {
+            // not enough memory to support the size!
+            if( mempool->stack.base - ALLOC_SIZE < mempool->stack.mem )
+                return NULL;
+            else {
+                // couldn't allocate from a freelist, allocate from available mempool.
+                // subtract allocation size from the mempool.
+                mempool->stack.base -= ALLOC_SIZE;
+                
+                // use the available mempool space as the new node.
+                new_mem = (struct MemNode *)mempool->stack.base;
```

Comment:
Potential non-aligned address stored. Results in undefined behavior here and in successive lines.

Reply:
the `malloc`d memory is aligned to the size of a `struct MemNode *`.
I've tested this memory pool on a Raspberry Pi 2 which is ARM based, there was no misalignment errors which triggers a bus error on ARM processors.

Reply:
Confirmed, no misalignment or bus errors on my Pi 2. If you have an ARM processor, I suggest testing this on that.